052-record-reconciled-versions.md

```diff
@@ -0,0 +1,190 @@
+
+## Motivation
+
+This proposal allows a user to rely on the `Kafka` custom resources as their single source of truth, instead of watching/querying the individual kafka/zookeeper pods.
+This proposal lays out all the information that is required, and at what points in the reconcile this information will be used. This information is already readily available within the Reconcile loop, so orchestrating and figuring out if a component has been fully upgraded will be a case of propagating this information into the Custom Resource via status fields.
+
+## Proposal
+
+This proposal will cover the `Kafka` custom resource but could apply to other CR types that use `StrimziPodSets` such as `KafkaConnect` and `KafkaMirrorMaker2`, but could be extended to include other Custom Resources if there would be value in it.
+This proposal starts with the `Kafka` Custom Resource as the base example as it is the core component of Strimzi, and it deploys both `StrimziPodSet`s and `Deployment`s.
+
+This proposal assumes that all key value pairs mentioned will be writen to the `status` field as this is managed by the operator and can be added to without impacting end users.
+
+The following are the proposed new fields:
+
+### status.operatorLastSuccessfulVersion
+
+The `status.operatorLastSuccessfulVersion` field is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion, meaning it reached the end of a reconcile without error.
+This would be the field a user would query against to check whether an operator upgrade was complete - or at least a complete roll of the Kafka brokers on the latest operator version. Details on how to verify that an upgrade has completed when the `Kafka.spec.kafka.version` is updated is covered in the next section.
+
+Example field in status:
+```
+status:
+  operatorLastSuccessfulVersion: '0.37.0'
+```
+If a reconcile does not reach completion, this field is not added or updated, however in the case where it already exists it is left at the previous value.
+i.e.
+`status.operatorLastSuccessfulVersion=0.35.0` would remain if a post-operator-upgrade reconciliation on `0.36.0` failed, as the information that a prior reconciliation passed would be useful to an end user.
+
+### status.kafkaVersion
+
+The `status.kafkaVersion` field is updated in the CR at the end of the KafkaReconciler in the `podsReady` method and signals to a user that the KafkaReconciler has finished reconciling the Kafka brokers to Kafka version `Y`, regardless of success or failure later in the reconcile.
+```
+status:
+  kafkaVersion '3.5.1'
+```
+
+Note: This will update the field in the reconcilers internal status object and have it be written at the end of the reconciliation rather then having a separate status update API call each time this is updated.
```

Comment:
```suggestion
Note: This will update the field in the reconciler's internal status object and have it be written at the end of the reconciliation rather then having a separate status update API call each time this is updated.
```